File: RailgunVB/Commands/JoinLeave/JoinLeaveAdd.vb
Imports Discord
Imports Discord.Commands
Imports TreeDiagram
Imports TreeDiagram.Enums
Imports TreeDiagram.Models.Server

Namespace Commands.JoinLeave
    
    Partial Public Class JoinLeave
    
        <Group("add")>
        Public Class JoinLeaveAdd
            Inherits ModuleBase
            
            Private ReadOnly _dbContext As TreeDiagramContext
    
            Public Sub New(dbContext As TreeDiagramContext)
                _dbContext = dbContext
            End Sub
            
            <Command("joinmsg")>
            Public Async Function JoinAsync(<Remainder> msg As String) As Task
                Await MsgHandlerAsync(msg, MsgType.Join)
            End Function
            
            <Command("leavemsg")>
            Public Async Function LeaveAsync(<Remainder> msg As String) As Task
                Await MsgHandlerAsync(msg, MsgType.Leave)
            End Function
            
            Private Async Function MsgHandlerAsync(msg As String, type As MsgType) As Task
                If String.IsNullOrWhiteSpace(msg)
                    await ReplyAsync("Please specify a message to add.")
                    Return
                End If
                
                Dim data As ServerJoinLeave = Await _dbContext.ServerJoinLeaves.GetOrCreateAsync(Context.Guild.Id)
                
                If (type = MsgType.Join AndAlso data.JoinMessages.Contains(msg)) OrElse 
                   (type = MsgType.Leave AndAlso data.LeaveMessages.Contains(msg))
                    await ReplyAsync("Specified message is already listed.")
                    Return
                End If
                
                data.AddMessage(msg, type)
                
                Await _dbContext.SaveChangesAsync()
                await ReplyAsync($"Successfully added {Format.Code(msg)} to {Format.Bold(type.ToString())} message.")
            End Function
        End Class
        
    End Class
    
End NameSpace

File: Railgun/Commands/JoinLeave/JoinLeaveAdd.cs
using System.Threading.Tasks;
using Discord;
using Finite.Commands;
using Railgun.Core;
using Railgun.Core.Attributes;
using TreeDiagram;
using TreeDiagram.Enums;
using TreeDiagram.Models;
using TreeDiagram.Models.Server;

namespace Railgun.Commands.JoinLeave
{
	public partial class JoinLeave : SystemBase
	{
		[Alias("add", "+")]
		public class JoinLeaveAdd : SystemBase
		{
			private ServerJoinLeave GetData(ulong guildId, bool create = false)
			{
				ServerProfile data;

				if (create)
					data = Context.Database.ServerProfiles.GetOrCreateData(guildId);
				else {
					data = Context.Database.ServerProfiles.GetData(guildId);

					if (data == null) 
						return null;
				}

				if (data.JoinLeave == null)
					if (create)
						data.JoinLeave = new ServerJoinLeave();
				
				return data.JoinLeave;
			}

			private Task MsgHandlerAsync(string msg, MsgType type)
			{
				if (string.IsNullOrWhiteSpace(msg)) return ReplyAsync("Please specify a message to add.");

				var data = GetData(Context.Guild.Id, true);

				if ((type == MsgType.Join && data.JoinMessages.Contains(msg)) || (type == MsgType.Leave && data.LeaveMessages.Contains(msg)))
					return ReplyAsync("Specified message is already listed.");

				data.AddMessage(msg, type);
				return ReplyAsync($"Successfully added {Format.Code(msg)} to {Format.Bold(type.ToString())} message.");
			}

			[Command("joinmsg")]
			public Task JoinAsync([Remainder] string msg) => MsgHandlerAsync(msg, MsgType.Join);

			[Command("leavemsg")]
			public Task LeaveAsync([Remainder] string msg) => MsgHandlerAsync(msg, MsgType.Leave);
		}
	}
}